### PR TITLE
272 proposal xml mask for xml in json

### PR DESCRIPTION
**Pull request by youen**

### CODE REVIEW

File: cmd/pimo/main.go [hunk 1]
```diff
@@ -168,23 +168,14 @@ There is NO WARRANTY, to the extent permitted by law.`, version, commit, buildDa
...
-					result := make(map[string]string, len(m))
-					for k, v := range transformedData {
-						stringValue, ok := v.(string)
-						if !ok {
-							return nil, fmt.Errorf("Result is not a string")
+					newList, _ := pimo.XMLCallback(ctx, m)
+					// remove element/attribute return *remove as value
+					for k := range m {
+						if _, ok := newList[k]; !ok {
+							newList[k] = "*remove"
```
Comment (youen) [inline cmd/pimo/main.go:175]: please don't use magic value '*remove'

File: pkg/xml/xml.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	pipeline := model.NewPipeline(nil)
+	pipeline, _, err = model.BuildPipeline(pipeline, definition, caches, fns, "", "")
+	return MaskEngine{xPath, injectParent, pipeline}, err
+}
+
+// Mask choose the target attribute or tag value and apply masking configuration
+func (engine MaskEngine) Mask(e model.Entry, context ...model.Dictionary) (model.Entry, error) {
+	log.Info().Msg("Mask XML")
+
+	// Prepare date
```
Comment (youen) [inline pkg/xml/xml.go:55]:
```suggestion
	// Prepare data
```

File: pkg/xml/xml.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	stringXML, ok := e.(string)
+	if !ok {
+		return nil, fmt.Errorf("jsonpath content is not a string")
+	}
+	// Check entry is a valid XML
+	if !isXMLValid(stringXML) {
```
Comment (youen) [inline pkg/xml/xml.go:61]: Why use another xml parser ? xixo parser can detect bad XML

File: pkg/xml/xml.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		}
+
+		dictContext := entryContext.(model.Dictionary).Unordered()
+		for k, v := range dictContext {
+			input.Set(engine.injectParent+k, v)
+		}
+	}
+	// Apply masking
+	parser.RegisterMapCallback(engine.xPath, func(m map[string]string) (map[string]string, error) {
+		for k, v := range m {
```
Comment (youen) [inline pkg/xml/xml.go:85]: the callback function could be externalized

**Comment by youen** [comment]
on error on pipe definition the error is not catch and show as a output :

```
[23:24] cannot unmarshal string into Go struct field Definition.Masking of type bool
  20 |                 dateMin: "1970-01-01T00:00:00Z"
  21 |                 dateMax: "2020-01-01T00:00:00Z"
  22 |             - template: "{{index . \"date\"}}"
> 23 |             - remove : tru
                              ^
```

Example to reproduce with pimo play :

#c=G4UwTgzglg9gdgLgAQCICMKBQBbAhhAayjgHMFMkkBaJCEAGxAGMAXGMcyrpAKwngAOuFgAtkKJvBYg4LLNzyFO3JAA9s9ZSrVDR4uDGnztSAMRJRIJELAyWSGACMezewApcSACYw8xAJRIAO5Q9PRIjla4TEwgENCOjMFQohYiVigA+ihIwLhgULiJVnC42CAAdBQmxC6sAAr5duLZ1dqKRKRaKjR0jKzs3Sa8-HC6YqgAArgArqLsxsMdQybmACowSDN0SJ42drm49DNWxBYg2AL0wiDIAN53FUgA1Ei1ro22sgByZVavfHgAGkQABPAC+4Law3Ol2u0nEDwqmRYKUYkMWPVoDFcg2hw0BY2EExQXhumPa+AIEBW2hoYFwcC8ABEbrSYWTpABZYjiNAATgA7AAGKjCtBitBrYXChAyuXCgBaFJhlE5IC5uFU4gATMK9ZLJdLZfKZcr8SYaNI4WzUA9iF4QKokE8ADqkm7ujEWulIWzYGCgJDIFhgGZAA&i=N4KABGBEAuCW0BsCmkBcUC2BPMAjBA9gOZgB2B0SYATJADThQDGBplbaUAPOZWAIYBXaAAsCAJwC8AcgBSBEaTAARAkmkA+LgBN+lDQEYADAHpjJ6keoBmLid36AKiNgBnMG4FkKVAgDMwbDxCIgA6cNC7XiQNSBAAXyA

**Comment by youen** [comment]
why i can't use dot notation 
for exemple : 

```
            # To use a parent value in template: {{. + injectParentName + . +  jsonKey}}
            template: "{{._.title}}"
```

**Comment by youen** [comment]
> on error on pipe definition the error is not catch and show as a output :
> 
> ```
> [23:24] cannot unmarshal string into Go struct field Definition.Masking of type bool
>   20 |                 dateMin: "1970-01-01T00:00:00Z"
>   21 |                 dateMax: "2020-01-01T00:00:00Z"
>   22 |             - template: "{{index . \"date\"}}"
> > 23 |             - remove : tru
>                               ^
> ```
> 
> Example to reproduce with pimo play :
> 
> #c=G4UwTgzglg9gdgLgAQCICMKBQBbAhhAayjgHMFMkkBaJCEAGxAGMAXGMcyrpAKwngAOuFgAtkKJvBYg4LLNzyFO3JAA9s9ZSrVDR4uDGnztSAMRJRIJELAyWSGACMezewApcSACYw8xAJRIAO5Q9PRIjla4TEwgENCOjMFQohYiVigA+ihIwLhgULiJVnC42CAAdBQmxC6sAAr5duLZ1dqKRKRaKjR0jKzs3Sa8-HC6YqgAArgArqLsxsMdQybmACowSDN0SJ42drm49DNWxBYg2AL0wiDIAN53FUgA1Ei1ro22sgByZVavfHgAGkQABPAC+4Law3Ol2u0nEDwqmRYKUYkMWPVoDFcg2hw0BY2EExQXhumPa+AIEBW2hoYFwcC8ABEbrSYWTpABZYjiNAATgA7AAGKjCtBitBrYXChAyuXCgBaFJhlE5IC5uFU4gATMK9ZLJdLZfKZcr8SYaNI4WzUA9iF4QKokE8ADqkm7ujEWulIWzYGCgJDIFhgGZAA&i=N4KABGBEAuCW0BsCmkBcUC2BPMAjBA9gOZgB2B0SYATJADThQDGBplbaUAPOZWAIYBXaAAsCAJwC8AcgBSBEaTAARAkmkA+LgBN+lDQEYADAHpjJ6keoBmLid36AKiNgBnMG4FkKVAgDMwbDxCIgA6cNC7XiQNSBAAXyA

a working example with pipe  : [pimo play withe error catched](https://cgi-fr.github.io/pimo-play/#c=G4UwTgzglg9gdgLgAQCICMKBQEQgCbIAsATJgLYCGEA1lHAOYKZJIC0SOANiAMYAuMMExYikAYiR8AFlAhIADhWkKYdPnIFIKcLWDAUAnkhgAzBeAjwIzUUgBWluIunIUg+tqgAvJbDgQAOnkLKyxRShphWwkoPiQAdyhOTiQAIxAFKGC8SRhJKQyI2gZM4M46DLwQEzpYvzSQThh4m1F5LJAo23F8jMUwEDg4mFS7XjiACgokPBhKOgBKBKSU9K0eHhAIaFTuZeVpDJQAfRQkYAowKApdjLgKMhBW2zox-gAFS8G+V1Pn8KoxUY-1s7C440EXW6tgc8GcUlc90eYWhAMiINREhqjTwcjo0CqvXMkHgxlG4yQPG0DXWm22+BmUAG-E4Bgx0L4IDI8k4Sk6qAA3gLJLE9gEkRkAL6SlGosGNCFCdndWFOJQI1AQACuYAlstRSCKUINki5PL5riFIr4Yu1uoeUplyrYHAV-EhztEqvhri5FCS+tRRs9Igkh3OFE4WoyEAEAxy+KghPD-W+ZLecVktK2Oz20jAMC19CkqFO+W01A0eRTXyGSFe4z8IZYAzIMFASGQYE5mEwQA&i=N4KABGBED2BODmBDAdgSwF6IC6usgzpAFxgDa4EYollkAJtALaKrLFQDGTADigJ4A6Lo0gAaCjUjcAprHx5CJcjRrUVk5IkbT2kADbS8iMRPVR8AV1ibtuxqj0HYJs7WnMHuyKZoBfca5qrpA2OiSQAFbQOgGu5lahunSIAG6ohLHB7ix6Xj6UvvkAuqb+pkG0DB5s4cK8yIIAZs6ZtDJyCuzKZhXqIVphUIh6LGytfZbWA3ayHKiyLnGQ2Z7h3q5lgfka0+GNenCoKIvBk4nhBvCwKHQnZsseuWvbYIXqJX4UJb5AA)

File: cmd/pimo/main.go (hunk 1 above)
Comment (Chao-Ma5566) [reply cmd/pimo/main.go:175]: sorry, I forgot to delete this in main.

**Comment by Chao-Ma5566** [comment]
> why i can't use dot notation for exemple :
> 
> ```
>             # To use a parent value in template: {{. + injectParentName + . +  jsonKey}}
>             template: "{{._.title}}"
> ```

Because Xixo only accept string value for the moment, so we can't add a dictionary in argument. Then We can't use this format in template, I did a little different like this `{{. + injectParentName  jsonKey}}`,  so it will be` template: "{{._title}}"`

**Review by Chao-Ma5566 — APPROVED** [review]

**Pull request merged**